File: test/functional/test_framework/chronik/alp.py
# Copyright (c) 2024 The Bitcoin developers
# Distributed under the MIT software license, see the accompanying
# file COPYING or http://www.opensource.org/licenses/mit-license.php.

from typing import List

from test_framework.messages import CTxOut
from test_framework.script import OP_RESERVED, OP_RETURN, CScript

# ALP integers are encoded as 48-bit little-endian integers
ALP_INT_SIZE = 6


def alp_opreturn(*sections: bytes) -> CTxOut:
    return CTxOut(0, CScript([OP_RETURN, OP_RESERVED] + list(sections)))


def extend_var_bytes(target: bytearray, b: bytes):
    target.append(len(b))
    target.extend(b)


def alp_genesis(
    *,
    token_ticker: bytes = b"",
    token_name: bytes = b"",
    url: bytes = b"",
    data: bytes = b"",
    auth_pubkey: bytes = b"",
    decimals: int = 0,
    mint_amounts: List[int],
    num_batons: int,
) -> bytes:
    result = bytearray()
    result.extend(b"SLP2")
    result.append(0)

    extend_var_bytes(result, b"GENESIS")
    extend_var_bytes(result, token_ticker)
    extend_var_bytes(result, token_name)
    extend_var_bytes(result, url)
    extend_var_bytes(result, data)
    extend_var_bytes(result, auth_pubkey)

    result.append(decimals)

    result.append(len(mint_amounts))
    for amount in mint_amounts:
        result.extend(amount.to_bytes(ALP_INT_SIZE, "little"))

    result.append(num_batons)
    return result


def alp_mint(
    token_id: str,
    mint_amounts: List[int],
    num_batons: int,
) -> bytes:
    result = bytearray()
    result.extend(b"SLP2")
    result.append(0)

    extend_var_bytes(result, b"MINT")

    result.extend(bytes.fromhex(token_id)[::-1])

    result.append(len(mint_amounts))
    for amount in mint_amounts:
        result.extend(amount.to_bytes(ALP_INT_SIZE, "little"))

    result.append(num_batons)

    return result


def alp_send(
    token_id: str,
    output_amounts: List[int],
) -> bytes:
    result = bytearray()
    result.extend(b"SLP2")
    result.append(0)

    extend_var_bytes(result, b"SEND")

    result.extend(bytes.fromhex(token_id)[::-1])

    result.append(len(output_amounts))
    for amount in output_amounts:
        result.extend(amount.to_bytes(ALP_INT_SIZE, "little"))

    return result


def alp_burn(
    token_id: str,
    burn_amount: int,
) -> bytes:
    result = bytearray()
    result.extend(b"SLP2")
    result.append(0)

    extend_var_bytes(result, b"BURN")

    result.extend(bytes.fromhex(token_id)[::-1])
    result.extend(burn_amount.to_bytes(ALP_INT_SIZE, "little"))

    return result

File: modules/ecash-lib/src/token/alp.ts
// Copyright (c) 2024 The Bitcoin developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

import { fromHex, fromHexRev } from '../io/hex.js';
import { strToBytes } from '../io/str.js';
import { Writer } from '../io/writer.js';
import { WriterBytes } from '../io/writerbytes.js';
import { WriterLength } from '../io/writerlength.js';
import { BURN, GENESIS, GenesisInfo, MINT, SEND } from './common.js';

/** LOKAD ID for ALP as string */
export const ALP_LOKAD_ID_STR = 'SLP2';

/** LOKAD ID for ALP */
export const ALP_LOKAD_ID = strToBytes(ALP_LOKAD_ID_STR);

/** ALP standard token type number */
export const ALP_STANDARD = 0;

export type AlpTokenType_Number = typeof ALP_STANDARD;

/** Supported ALP token types */
export interface AlpTokenType {
    protocol: 'ALP';
    type: AlpTokenType_Type;
    number: number;
}
/** Possible ALP token types returned by chronik */
export type AlpTokenType_Type =
    | 'ALP_TOKEN_TYPE_STANDARD'
    | 'ALP_TOKEN_TYPE_UNKNOWN';

/** ALP limits lengths/sizes to this number, e.g. the number of outputs */
export const ALP_MAX_SIZE = 127;

export const ALP_TOKEN_TYPE_STANDARD: AlpTokenType = {
    protocol: 'ALP',
    type: 'ALP_TOKEN_TYPE_STANDARD',
    number: ALP_STANDARD,
};

/**
 * Although ALP_MAX_SIZE is 127, in practice we can only
 * handle 29 ALP outputs in a single OP_RETURN given the curent
 * 223-byte OP_RETURN size limit, and even this assumes
 * we are only working with 1 ALP token.
 *
 * For example an ALP tx that sends multiple tokens cannot support
 * 29 token outputs as the instructions will require more than 223
 * bytes in the OP_RETURN.
 *
 * So, this is an upper bound on ALP outputs per current mempool
 * acceptance rules
 */
export const ALP_POLICY_MAX_OUTPUTS = 29;

/** Mint data specifying mint amounts (in atoms) and batons of a GENESIS/MINT tx */
export interface MintData {
    /**
     * Array of amounts in atoms to be minted by this tx, each having their own tx output.
     */
    atomsArray: bigint[];
    /** Number of mint batons to create, each having their own tx output. */
    numBatons: number;
}

/** Build an ALP GENESIS pushdata section, creating a new ALP token */
export function alpGenesis(
    tokenType: number,
    genesisInfo: GenesisInfo,
    mintData: MintData,
): Uint8Array {
    const writeSection = (writer: Writer) => {
        writer.putBytes(ALP_LOKAD_ID);
        writer.putU8(tokenType);
        putVarBytes(GENESIS, writer);
        putVarBytes(strToBytes(genesisInfo.tokenTicker ?? ''), writer);
        putVarBytes(strToBytes(genesisInfo.tokenName ?? ''), writer);
        putVarBytes(strToBytes(genesisInfo.url ?? ''), writer);
        putVarBytes(fromHex(genesisInfo.data ?? ''), writer);
        putVarBytes(fromHex(genesisInfo.authPubkey ?? ''), writer);
        writer.putU8(genesisInfo.decimals ?? 0);
        putMintData(mintData, writer);
    };
    const writerLength = new WriterLength();
    writeSection(writerLength);
    const writerBytes = new WriterBytes(writerLength.length);
    writeSection(writerBytes);
    return writerBytes.data;
}

/**
 * Build an ALP MINT pushdata section, creating new ALP tokens and mint batons
 * of the given token ID.
 **/
export function alpMint(
    tokenId: string,
    tokenType: number,
    mintData: MintData,
): Uint8Array {
    const tokenIdBytes = fromHexRev(tokenId);
    const writeSection = (writer: Writer) => {
        writer.putBytes(ALP_LOKAD_ID);
        writer.putU8(tokenType);
        putVarBytes(MINT, writer);
        writer.putBytes(tokenIdBytes);
        putMintData(mintData, writer);
    };
    const writerLength = new WriterLength();
    writeSection(writerLength);
    const writerBytes = new WriterBytes(writerLength.length);
    writeSection(writerBytes);
    return writerBytes.data;
}

/**
 * Build an ALP SEND pushdata section, moving ALP tokens to different outputs
 **/
export function alpSend(
    tokenId: string,
    tokenType: number,
    sendAtomsArray: bigint[],
): Uint8Array {
    const tokenIdBytes = fromHexRev(tokenId);
    const writeSection = (writer: Writer) => {
        writer.putBytes(ALP_LOKAD_ID);
        writer.putU8(tokenType);
        writer.putU8(SEND.length);
        writer.putBytes(SEND);
        writer.putBytes(tokenIdBytes);
        writer.putU8(sendAtomsArray.length);
        for (const atoms of sendAtomsArray) {
            putAlpAtoms(atoms, writer);
        }
    };
    const writerLength = new WriterLength();
    writeSection(writerLength);
    const writerBytes = new WriterBytes(writerLength.length);
    writeSection(writerBytes);
    return writerBytes.data;
}

/** Build an ALP BURN pushdata section, intentionally burning ALP tokens. */
export function alpBurn(
    tokenId: string,
    tokenType: number,
    burnAtoms: bigint,
): Uint8Array {
    const tokenIdBytes = fromHexRev(tokenId);
    const writeSection = (writer: Writer) => {
        writer.putBytes(ALP_LOKAD_ID);
        writer.putU8(tokenType);
        writer.putU8(BURN.length);
        writer.putBytes(BURN);
        writer.putBytes(tokenIdBytes);
        putAlpAtoms(burnAtoms, writer);
    };
    const writerLength = new WriterLength();
    writeSection(writerLength);
    const writerBytes = new WriterBytes(writerLength.length);
    writeSection(writerBytes);
    return writerBytes.data;
}

function putMintData(mintData: MintData, writer: Writer) {
    writer.putU8(mintData.atomsArray.length);
    for (const atoms of mintData.atomsArray) {
        putAlpAtoms(atoms, writer);
    }
    writer.putU8(mintData.numBatons);
}

function putAlpAtoms(atoms: bigint, writer: Writer) {
    const atomsN = BigInt(atoms);
    writer.putU32(atomsN & 0xffffffffn);
    writer.putU16(atomsN >> 32n);
}

function putVarBytes(bytes: Uint8Array, writer: Writer) {
    if (bytes.length > 127) {
        throw new Error('Length of bytes must be between 0 and 127');
    }
    writer.putU8(bytes.length);
    writer.putBytes(bytes);
}
